8316930: HotSpot should use noexcept instead of throw()

TheShermanTanker opened the pull request:
throw() has been deprecated since C++11 alongside dynamic exception specifications, we should replace all instances of it with noexcept to prepare HotSpot for later versions of C++

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warnings
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/nfc.nrm)
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/nfkc.nrm)
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/ubidi.icu)
&nbsp;⚠️ Patch contains a binary file (src/java.base/share/classes/jdk/internal/icu/impl/data/icudt74b/uprops.icu)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/text/doc-files/Document-insert.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw16.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw24.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw32.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-bw48.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim16.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim24.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim32.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-interim48.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow16.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow24.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow32.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/sun/awt/resources/security-icon-yellow48.png)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/windows/native/libawt/windows/security_warning.ico)
&nbsp;⚠️ Patch contains a binary file (src/utils/IdealGraphVisualizer/View/src/main/resources/com/sun/hotspot/igv/view/images/hideDuplicates.png)
&nbsp;⚠️ Patch contains a binary file (test/jdk/java/security/ProtectionDomain/AllPerm.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TokenStore.keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TrustedCert.keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyFile/TrustedCert.keystore1)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyParser/ExtDirsA/a.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/security/provider/PolicyParser/ExtDirsB/b.jar)

### Issue
 * [JDK-8316930](https://bugs.openjdk.org/browse/JDK-8316930): HotSpot should use noexcept instead of throw() (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15910/head:pull/15910` \
`$ git checkout pull/15910`

Update a local copy of the PR: \
`$ git checkout pull/15910` \
`$ git pull https://git.openjdk.org/jdk.git pull/15910/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15910`

View PR using the GUI difftool: \
`$ git pr show -t 15910`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15910.diff">https://git.openjdk.org/jdk/pull/15910.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15910#issuecomment-1734751964)
</details>

dholmes-ora reviewed (commented):
The change from `throw()` to `noexcept` seems reasonable though I assume we have to first approve this via the Hostpot style guide?

There are places you are adding `noexcept` when there is no `throw()` and that seems inappropriate for this PR - unless they are required because of some transitive rule application?

Thanks.

TheShermanTanker commented:
> The change from `throw()` to `noexcept` seems reasonable though I assume we have to first approve this via the Hostpot style guide?
> 
> There are places you are adding `noexcept` when there is no `throw()` and that seems inappropriate for this PR - unless they are required because of some transitive rule application?
> 
> Thanks.

Oops, those places where noexcept slipped through are from an experimental branch where I enabled exceptions for HotSpot, they got in by mistake, my apologies. I've moved this to draft for now, since Apple Clang seems to be having a hard time with the noexcept specifiers. Not sure about the Style Guide, maybe I should ask @kbarrett?

dholmes-ora commented:
I will ping @kimbarrett internally.

kimbarrett commented:
> > The change from `throw()` to `noexcept` seems reasonable though I assume we have to first approve this via the Hostpot style guide?
> > There are places you are adding `noexcept` when there is no `throw()` and that seems inappropriate for this PR - unless they are required because of some transitive rule application?
> > Thanks.
> 
> Oops, those places where noexcept slipped through are from an experimental branch where I enabled exceptions for HotSpot, they got in by mistake, my apologies. I've moved this to draft for now, since Apple Clang seems to be having a hard time with the noexcept specifiers. Not sure about the Style Guide, maybe I should ask @kimbarrett?

I've added https://bugs.openjdk.org/browse/JDK-8255082 as a blocker for this bug.  As noted there, addressing that
is not as simple as just saying "yes, we can use noexcept".

And no, we're not going to enable exceptions in HotSpot.

I'd be interested in hearing what problems you've run into with Apple clang wrto noexcept.

TheShermanTanker commented:
> > > The change from `throw()` to `noexcept` seems reasonable though I assume we have to first approve this via the Hostpot style guide?
> > > There are places you are adding `noexcept` when there is no `throw()` and that seems inappropriate for this PR - unless they are required because of some transitive rule application?
> > > Thanks.
> > 
> > 
> > Oops, those places where noexcept slipped through are from an experimental branch where I enabled exceptions for HotSpot, they got in by mistake, my apologies. I've moved this to draft for now, since Apple Clang seems to be having a hard time with the noexcept specifiers. Not sure about the Style Guide, maybe I should ask @kimbarrett?
> 
> I've added https://bugs.openjdk.org/browse/JDK-8255082 as a blocker for this bug. As noted there, addressing that is not as simple as just saying "yes, we can use noexcept".
> 
> And no, we're not going to enable exceptions in HotSpot.
> 
> I'd be interested in hearing what problems you've run into with Apple clang wrto noexcept.

No worries, the branch with the Exceptions enabled HotSpot is for my own personal use. Apple Clang seems to have issues parsing noexcept in operator new, as seen in the tests for this PR:

```
make[1]: *** [/Users/runner/work/jdk/jdk/make/Init.gmk:323: main] Error 2
=== Output from failing command(s) repeated here ===
make: *** [/Users/runner/work/jdk/jdk/make/Init.gmk:[189](https://github.com/TheShermanTanker/jdk/actions/runs/6307800420/job/17125236356#step:8:191): product-bundles] Error 2
* For target hotspot_variant-server_tools_adlc_objs_adlArena.o:
In file included from /Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.cpp:25:
In file included from /Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlc.hpp:88:
/Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.hpp:43:34: error: expected ';' at end of declaration list
  void* operator new(size_t size) noexcept;
                                 ^
                                 ;
/Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.hpp:52:34: error: expected ';' at end of declaration list
  void* operator new(size_t size) noexcept;
                                 ^
                                 ;
/Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.hpp:66:49: error: expected ';' at end of declaration list
  void* operator new(size_t size, size_t length) noexcept;
                                                ^
                                                ;
/Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.cpp:47:68: error: expected function body after function declarator
   ... (rest of output omitted)
* For target hotspot_variant-server_tools_adlc_objs_adlparse.o:
In file included from /Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlparse.cpp:27:
In file included from /Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlc.hpp:88:
/Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.hpp:43:34: error: expected ';' at end of declaration list
  void* operator new(size_t size) noexcept;
                                 ^
                                 ;
/Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.hpp:52:34: error: expected ';' at end of declaration list
  void* operator new(size_t size) noexcept;
                                 ^
                                 ;
/Users/runner/work/jdk/jdk/src/hotspot/share/adlc/adlArena.hpp:66:49: error: expected ';' at end of declaration list
  void* operator new(size_t size, size_t length) noexcept;
                                                ^
                                                ;
3 errors generated.
```

TheShermanTanker commented:
Something seems to be wrong with Apple Clang here, hmm...

TheShermanTanker commented on src/hotspot/share/c1/c1_Instruction.hpp:
```diff
@@ -334,7 +334,7 @@ class Instruction: public CompilationResourceObj {
 
  public:
-  void* operator new(size_t size) throw() {
+  void* operator new(size_t size) noexcept {
```
Should the noexcept specifiers on both operator new in this file be removed? They both call Amalloc with the default parameter of EXIT\_OOM, which means they both crash instead of returning nullptr when failing to allocate memory

dholmes-ora replied:
I think EXIT_OOM is an implementation detail. The key point is that these methods do not throw.

dholmes-ora replied:
Or do I have that backwards ... the key point of `no_except` is that callers of these method must check for `nullptr`, but with EXIT_OOM that is not the case - and we don't want it to appear that the allocation can actually fail and we continue execution!

kimbarrett replied:
It looks like there are a number of `operator new`s that have nothrow exception specs but shouldn't.
CompilationResourceObj in the immediately preceding file, for example.
A precursor cleanup (or maybe several) that removed those first would be nice.

kimbarrett replied:
It seems that https://bugs.openjdk.org/browse/JDK-8305590 only removed some of them.

TheShermanTanker commented:
/open

TheShermanTanker commented:
I would like to address this soon, but will probably need help writing noexcept for the Style Guide

TheShermanTanker commented:
/open
That was very rude of you

TheShermanTanker commented:
You're high Skara, go home
